reactos/drivers/usb/usbport/device.c

```diff
@@ -0,0 +1,1873 @@
+
+        InterfaceHandle = (PUSBPORT_INTERFACE_HANDLE)ExAllocatePoolWithTag(NonPagedPool,
+                                                                           HandleLength,
+                                                                           USB_PORT_TAG);
+
+        if (!InterfaceHandle)
+        {
+            USBDStatus = USBD_STATUS_INSUFFICIENT_RESOURCES;
+            goto Exit;
+        }
+
+        RtlZeroMemory(InterfaceHandle, HandleLength);
+
+        ix = NumInterfaces;
+
+        if (NumInterfaces > 0)
+        {
+            PipeHandle = &InterfaceHandle->PipeHandle[0];
+
+            do
+            {
+                PipeHandle->Flags = PIPE_HANDLE_FLAG_CLOSED;
+                PipeHandle->Endpoint = NULL;
+
+                PipeHandle += 1;
+                --ix;
+            }
+            while (ix);
+        }
+
+        IsAllocated = TRUE;
+    }
+
+    InterfaceHandle->AlternateSetting = InterfaceInfo->AlternateSetting;
+
+    RtlCopyMemory(&InterfaceHandle->InterfaceDescriptor,
+                  InterfaceDescriptor,
+                  sizeof(USB_INTERFACE_DESCRIPTOR));
+
+    InterfaceInfo->Class = InterfaceDescriptor->bInterfaceClass;
+    InterfaceInfo->SubClass = InterfaceDescriptor->bInterfaceSubClass;
+    InterfaceInfo->Protocol = InterfaceDescriptor->bInterfaceProtocol;
+    InterfaceInfo->Reserved = 0;
+    InterfaceInfo->NumberOfPipes = InterfaceDescriptor->bNumEndpoints;
+
+    Descriptor = (PUSB_ENDPOINT_DESCRIPTOR)((ULONG_PTR)InterfaceDescriptor +
+                                            InterfaceDescriptor->bLength);
+
+    ix = 0;
+
+    if (NumInterfaces)
+    {
+        PipeHandle = &InterfaceHandle->PipeHandle[0];
+
+        do
+        {
+            if (Descriptor->bDescriptorType != USB_ENDPOINT_DESCRIPTOR_TYPE)
+            {
+                do
+                {
+                    if (Descriptor->bLength == 0)
+                        break;
+                    else
+                        Descriptor = (PUSB_ENDPOINT_DESCRIPTOR)((ULONG_PTR)Descriptor +
+                                                                Descriptor->bLength);
+                }
+                while (Descriptor->bDescriptorType != USB_ENDPOINT_DESCRIPTOR_TYPE);
+            }
+
+            if (InterfaceInfo->Pipes[ix].PipeFlags & USBD_PF_CHANGE_MAX_PACKET)
+                Descriptor->wMaxPacketSize = InterfaceInfo->Pipes[ix].MaximumPacketSize;
+
+            RtlCopyMemory(&PipeHandle->EndpointDescriptor,
+                          Descriptor,
+                          sizeof(USB_ENDPOINT_DESCRIPTOR));
+
+            PipeHandle->Flags = PIPE_HANDLE_FLAG_CLOSED;
+            PipeHandle->PipeFlags = InterfaceInfo->Pipes[ix].PipeFlags;
+            PipeHandle->Endpoint = NULL;
+
+            wMaxPacketSize = Descriptor->wMaxPacketSize;
+
+            MaxPacketSize = (wMaxPacketSize & 0x7FF) * (((wMaxPacketSize >> 11) & 3) + 1);
```

Comment:
It for High-Speed devices.  Endpoints may have 1, 2 or 3 transactions per microframe.
USB 2.0 Specification, 5.9  High-Speed, High Bandwidth Endpoints